src/lib/tls/tls13_pqc/hybrid_public_key.cpp

```diff
@@ -0,0 +1,373 @@
+/**
+* Composite key pair that exposes the Public/Private key API but combines
+* multiple key agreement schemes into a hybrid algorithm.
+*
+* (C) 2023 Jack Lloyd
+*     2023 Fabian Albert, René Meusel - Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+*/
+
+#include <botan/internal/hybrid_public_key.h>
+
+#include <botan/pk_algs.h>
+
+#include <botan/internal/fmt.h>
+#include <botan/internal/kex_to_kem_adapter.h>
+#include <botan/internal/pk_ops_impl.h>
+#include <botan/internal/stl_util.h>
+
+namespace Botan::TLS {
+
+namespace {
+
+template <typename RetT, typename KeyT, typename ReducerT>
+RetT reduce(const std::vector<KeyT>& keys, RetT acc, ReducerT reducer) {
+   for(const KeyT& key : keys) {
+      acc = reducer(std::move(acc), key);
+   }
+   return acc;
+}
+
+std::vector<std::pair<std::string, std::string>> algorithm_specs_for_group(Group_Params group) {
+   BOTAN_ASSERT_NOMSG(is_hybrid(group));
+
+   switch(group) {
+      case Group_Params::HYBRID_X25519_KYBER_512_R3_OQS:
+      case Group_Params::HYBRID_X25519_KYBER_512_R3_CLOUDFLARE:
+         return {{"Curve25519", "Curve25519"}, {"Kyber", "Kyber-512-r3"}};
+      case Group_Params::HYBRID_X25519_KYBER_768_R3_OQS:
+      case Group_Params::HYBRID_X25519_KYBER_768_R3_CLOUDFLARE:
+         return {{"Curve25519", "Curve25519"}, {"Kyber", "Kyber-768-r3"}};
+
+      case Group_Params::HYBRID_SECP256R1_KYBER_512_R3_OQS:
+         return {{"ECDH", "secp256r1"}, {"Kyber", "Kyber-512-r3"}};
+      case Group_Params::HYBRID_SECP384R1_KYBER_768_R3_OQS:
+         return {{"ECDH", "secp384r1"}, {"Kyber", "Kyber-768-r3"}};
+      case Group_Params::HYBRID_SECP521R1_KYBER_1024_R3_OQS:
+         return {{"ECDH", "secp521r1"}, {"Kyber", "Kyber-1024-r3"}};
+
+      default:
+         return {};
+   }
+}
+
+std::vector<AlgorithmIdentifier> algorithm_identifiers_for_group(Group_Params group) {
+   BOTAN_ASSERT_NOMSG(is_hybrid(group));
+
+   const auto specs = algorithm_specs_for_group(group);
+   std::vector<AlgorithmIdentifier> result;
+
+   // This maps the string-based algorithm specs hard-coded above to OID-based
+   // AlgorithmIdentifier objects. The mapping is needed because
+   // load_public_key() depends on those while create_private_key() requires the
+   // strong-based spec.
+   //
+   // TODO: This is inconvenient, confusing and error-prone. Find a better way
+   //       to load arbitrary public keys.
```

Comment:
I agree this is awkward. Can you open a ticket about this?

Reply:
https://github.com/randombit/botan/issues/3706